source/extensions/common/redis/redirection_mgr_impl.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <array>
+#include <atomic>
+#include <numeric>
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "common/common/lock_guard.h"
+#include "common/common/thread.h"
+
+#include "extensions/common/redis/redirection_mgr.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Redis {
+
+class RedirectionManagerImpl : public RedirectionManager,
+                               public Envoy::Singleton::Instance,
+                               public std::enable_shared_from_this<RedirectionManagerImpl> {
+public:
+  friend class RedirectionMgrTest;
+
+  /**
+   * The information that the manager keeps for each cluster upon registration.
+   */
+  struct ClusterInfo {
+    ClusterInfo(std::string cluster_name, std::chrono::milliseconds min_time_between_triggering,
+                uint32_t redirects_threshold, RedirectCB cb)
+        : cluster_name_(std::move(cluster_name)),
+          min_time_between_triggering_(min_time_between_triggering),
+          redirects_threshold_(redirects_threshold), cb_(std::move(cb)) {}
+    std::string cluster_name_;
+    std::atomic<uint64_t> last_callback_time_ms_{};
+    std::atomic<uint32_t> redirects_count_{};
+    std::chrono::milliseconds min_time_between_triggering_;
+    uint32_t redirects_threshold_;
+    RedirectCB cb_;
+  };
+
+  using ClusterInfoSharedPtr = std::shared_ptr<ClusterInfo>;
+
+  class HandleImpl : public Handle {
+  public:
+    HandleImpl(RedirectionManagerImpl* mgr, ClusterInfoSharedPtr& cluster_info)
+        : manager_(mgr->shared_from_this()), cluster_info_(cluster_info) {}
+
+    ~HandleImpl() override {
+      std::shared_ptr<RedirectionManagerImpl> strong_manager = manager_.lock();
+      ClusterInfoSharedPtr strong_cluster_info = cluster_info_.lock();
+      if (strong_manager && strong_cluster_info) {
```

Comment:
Can this happen or should this be an ASSERT? I'm not sure if the manager vs. lifetime issues here.

Reply:
At this point I can simplify this greatly by holding onto shared pointers instead of weak pointers in the handle object.